lib/ex_doc/language/source.ex

```diff
@@ -0,0 +1,261 @@
+    filtermap_ast(abst_code, nil, fn
+      {:attribute, anno, kind, {mfa, ast}} when kind in [:spec, :callback] ->
+        ast = Enum.map(ast, &expand_records(&1, records))
+        {:attribute, anno, kind, {mfa, ast}}
+
+      {:attribute, anno, type, {name, ast, args}} when type in [:opaque, :type] ->
+        {:attribute, anno, type, {name, expand_records(ast, records), args}}
+
+      otherwise ->
+        otherwise
+    end)
+  end
+
+  defp expand_records(types, records) when is_list(types) do
+    Enum.map(types, &expand_records(&1, records))
+  end
+
+  defp expand_records({:ann_type, anno, [name, type]}, records) do
+    {:ann_type, anno, [name, expand_records(type, records)]}
+  end
+
+  ## When we encounter a record, we fetch the type definitions in the record and
+  ## merge then with the type. If there are duplicates we take the one in the type
+  ## declaration
+  defp expand_records({:type, anno, :record, [{:atom, _, record} = name | args]}, records) do
+    args =
+      (args ++ Map.get(records, record, []))
+      |> Enum.uniq_by(fn {:type, _, :field_type, [{:atom, _, name} | _]} -> name end)
+
+    ## We delete the record from the map so that recursive
+    ## record definitions are not expanded.
+    records = Map.delete(records, record)
+
+    {:type, anno, :record, expand_records([name | args], records)}
+  end
+
+  defp expand_records({type, anno, what, args}, records) when type in [:type, :user_type] do
+    {:type, anno, what, expand_records(args, records)}
+  end
+
+  defp expand_records({:remote_type, anno, [m, t, args]}, records) do
+    {:remote_type, anno, [m, t, expand_records(args, records)]}
+  end
+
+  defp expand_records(otherwise, _records) do
+    otherwise
+  end
+
+  @doc """
+  Get the basedir of a module
+
+  The basedir is the cwd of the Elixir/Erlang compiler when compiling the module.
+  All `-file` attributes in the module is relative to this directory.
+  """
+  def get_basedir(abst_code, module) do
+    ## We look for the first -file attribute to see what the source file that
+    ## was compiled is called.
+    filename =
+      Enum.find_value(abst_code, fn
+        {:attribute, _anno, :file, {filename, _line}} ->
+          filename
+
+        _ ->
+          nil
+      end)
+
+    ## The first -file attribute will be either relative or absolute
+    ## depending on whether the compiler was called with an absolute
+    ## or relative path.
+    if Path.type(filename) == :relative do
+      ## If the compiler was called with a relative path, then any other
+      ## relative -file attribute will be relative to the same directory.
+      ## We use `module_info(:compile)[:source]` to get an absolute path
+      ## to the source file and calculate the basedir from that
+
+      module.module_info(:compile)[:source]
+      |> String.Chars.to_string()
+      |> Path.absname()
+      |> Path.split()
+      |> Enum.reverse()
+      |> Enum.split(Path.split(filename) |> Enum.count())
+      |> elem(1)
```

Comment:
`drop` or `take` will give you the sides of `split`. I think here you want drop (please double check)

```suggestion
      |> Enum.drop(Path.split(filename) |> Enum.count())
```